File: ctci/chapter_1_arrays_strings/js/1.4-palindrome.js
/* O(n) solution */

/**
 * Determines if character is in the alphabet.
 * @param   {string}  ch
 * @returns {boolean}
 */
const isAlpha = ch => {
  /* case insensitive regex */
  return /^[A-Z]$/i.test(ch)
}

/**
 * Determines if string or its permutations can be a palindrome.
 * @param   {string}  str
 * @param   {boolean}
 */
const isPaliPermutation = str => {
  const obj = {}
  let odds = 0
  for (let ch of str) {
    if (isAlpha(ch)) {
      obj[ch] == undefined ? (obj[ch] = 1) : obj[ch]++
      obj[ch] % 2 != 0 ? odds++ : odds--
    }
  }
  /* Palindrome should have max one odd value at the end of loop. */
  return odds <= 1
}

console.log(isPaliPermutation("tact coa")) // true
console.log(isPaliPermutation("tact coao")) // true
console.log(isPaliPermutation("tact ccoa")) // false
console.log(isPaliPermutation("tact coaz")) // false
console.log(isPaliPermutation("at ta")) // true
console.log(isPaliPermutation("t")) // true

File: ctci/chapter_1_arrays_strings/python/1.4-palindrome.py
# O(n) solution


def is_pali_permutation(string: str) -> bool:
    """
    Determines if string or its permutations can be a palindrome.
    """
    _dict = {}
    odds = 0
    for ch in string:
        if ch.isalpha():
            if ch in _dict:
                _dict[ch] += 1
            else:
                _dict[ch] = 1
            if _dict[ch] % 2 != 0:
                odds += 1
            else:
                odds -= 1
    # Palindrome should have max one odd value at the end of loop.
    return odds <= 1


if __name__ == 'main':
    print(is_pali_permutation("tact coa"))  # true
    print(is_pali_permutation("tact coao"))  # true
    print(is_pali_permutation("tact ccoa"))  # false
    print(is_pali_permutation("tact coaz"))  # false
    print(is_pali_permutation("at ta"))  # true
    print(is_pali_permutation("t"))  # true
